servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/ZlibOpenSslCertificateCompressionAlgorithm.java

```diff
@@ -0,0 +1,100 @@
+
+import io.netty.handler.ssl.OpenSslCertificateCompressionAlgorithm;
+
+import java.io.ByteArrayOutputStream;
+import java.util.zip.Deflater;
+import java.util.zip.Inflater;
+import javax.net.ssl.SSLEngine;
+
+/**
+ * Implements ZLIB compression and decompression of OpenSSL certificates.
+ *
+ * @see <a href="https://www.zlib.net">ZLIB Website</a>
+ */
+final class ZlibOpenSslCertificateCompressionAlgorithm implements OpenSslCertificateCompressionAlgorithm {
+
+    @Override
+    public byte[] compress(final SSLEngine engine, final byte[] uncompressedCertificate) {
+        int uncompressedLength = uncompressedCertificate.length;
+        if (uncompressedLength == 0) {
+            return uncompressedCertificate;
+        }
+
+        final Deflater deflater = new Deflater();
+
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+            deflater.setInput(uncompressedCertificate);
+            deflater.finish();
+
+            // This calculation (which is also used inside Netty) comes from the C library which describes it as:
+            // "...must be at least 0.1% larger than the uncompressed length plus 12 bytes..."
+            int bufferSizeEstimate = (int) Math.ceil(uncompressedLength * 1.001) + 12;
+
+            byte[] compressionBuffer = new byte[bufferSizeEstimate];
+            while (!deflater.finished()) {
+                int bytesCompressed = deflater.deflate(compressionBuffer);
+                outputStream.write(compressionBuffer, 0, bytesCompressed);
+            }
+            return outputStream.toByteArray();
+        } catch (Exception cause) {
+            throw new CertificateCompressionException("Failed to compress certificate with " + getClass(), cause);
+        } finally {
+            deflater.end();
+        }
+    }
+
+    @Override
+    public byte[] decompress(final SSLEngine engine, final int uncompressedLen, final byte[] compressedCertificate) {
+        if (compressedCertificate.length == 0) {
+            return compressedCertificate;
+        }
+
+        final Inflater inflater = new Inflater();
+        try {
+            inflater.setInput(compressedCertificate);
+
+            // We do not need to create a ByteArrayOutputStream like we do on compression, since we know the maximum
+            // size on decompress is provided as an argument and anything larger would be a violation of the RFC so
+            // it will be rejected with an Exception.
+            byte[] output = new byte[uncompressedLen];
+            int bytesWritten = 0;
+            while (!inflater.finished()) {
+                int decompressedBytes = inflater.inflate(output, bytesWritten, uncompressedLen - bytesWritten);
+                bytesWritten += decompressedBytes;
+                if (bytesWritten > uncompressedLen) {
+                    throw new CertificateCompressionException("Number of bytes written exceeds the " +
+                            "uncompressed certificate length");
```

Comment:
Consider including values of `bytesWritten` and `uncompressedLen` in the exception message